### PR TITLE
[4.0] Remove required

### PR DESCRIPTION
**Pull request by brianteeman**
If you have a field with a default value then there is no point in also setting it as required as it already has a value

### CODE REVIEW

File: installation/forms/preinstall.xml [hunk 1]
```diff
@@ -18,7 +17,6 @@
 			label="INSTL_FTP_ENABLE_LABEL"
 			id="ftp_enable"
 			class="btn-group btn-group-reverse"
-			default="0"
```
Comment (Quy) [inline installation/forms/preinstall.xml]: Revert and remove `required`.

File: installation/forms/setup.xml [hunk 2]
```diff
@@ -51,7 +51,6 @@
 			class="custom-select form-control"
 			supported="mysql,mysqli,pgsql,postgresql"
 			required="true"
-			default="mysqli"
```
Comment (Quy) [inline installation/forms/setup.xml]: Revert and remove `required`.

File: installation/forms/setup.xml [hunk 3]
```diff
@@ -60,7 +59,6 @@
 			label="INSTL_DATABASE_HOST_DESC"
 			id="db_host"
 			class="form-control"
-			default="localhost"
```
Comment (Quy) [inline installation/forms/setup.xml]: Revert since it is a common value?

**Comment by brianteeman** [comment]
@quy thanks - not sure what I was thinking. I also put back the required on the text fields even if they have a default

File: plugins/authentication/cookie/cookie.xml [hunk 4]
```diff
@@ -25,7 +25,6 @@
 					label="PLG_AUTH_COOKIE_FIELD_COOKIE_LIFETIME_LABEL"
 					default="60"
 					filter="integer"
-					required="true"
```
Comment (Quy) [inline plugins/authentication/cookie/cookie.xml]: Revert per your last comment.

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 59fee06d60b0e26c863ac322268c46a92c91aec0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24982">issues.joomla.org/tracker/joomla-cms/24982</a>.</sub>

**Comment by brianteeman** [comment]
Updating branch to trigger drone

**Comment by wilsonge** [comment]
This is actually required. The serverside validation uses the required attribute to ensure the properties are present when submitting. Example rule with relevent lines: https://github.com/joomla/joomla-cms/blob/4.0-dev/libraries/src/Form/Rule/OptionsRule.php#L42-L48 (this stops people submitting forms through post requests etc with missing attributes)

**Comment by brianteeman** [comment]
Makes no sense to me to mark a field as required when it always has a value.

**Comment by brianteeman** [comment]
And there is no consistency or documentation

**Comment by HLeithner** [comment]
It makes sense, because you prevent the user to submit an empty field and get an unexpected value.

**Comment by wilsonge** [comment]
It's not just about the client side validation. It's the server side validation. You can send a post request to most these endpoints without ever using the client interface and we still need to protect against that

**Comment by HLeithner** [comment]
I think Brian expects that the default value is used in this case.

**Comment by brianteeman** [comment]
Exactly. And the complete lack of consistency negates everything else

**Comment by HLeithner** [comment]
And that's not true, if I want to remove the value of a field I expect that it's empty afterwards and doesn't have the default value.

Changing this would automatically set every empty field to it's default and that is not what you want.

**Comment by brianteeman** [comment]
I stand by my comments

**Comment by HLeithner** [comment]
So you say its better to make it impossible to submit an empty string? Sorry that's nonsense.

**Comment by brianteeman** [comment]
If we need the field to have a value then it should be impossible to submit an empty string. The fact that you can submit empty strings in all sorts of places when we must have a value is why we have errors such as #24972

**Comment by HLeithner** [comment]
if we need a value we use the require attribute that's the reason it exists.

**Comment by brianteeman** [comment]
and as shown that is not consistent and causes problems because of your "nonsense" belief that you can submit an empty string.

I have closed this PR - you can deal with the fallout

**Comment by mbabker** [comment]
> So you say its better to make it impossible to submit an empty string? Sorry that's nonsense.

If a field is required, generally that means it must have a non-empty value, therefore submitting the field with an empty string should throw a validation error.  Assuming you are able to misuse the required attribute of a form field, which has a semantic definition per the HTML standard, to communicate "this field must be submitted as part of the POST, but if it's empty that's perfectly OK" just proves you do not understand basic concepts of data validation.

Misusing the Form API to define structures for the webservices API and a validation layer is also architecturally incorrect, but as Joomla continues to be dominated by a bunch of amateurs who don't have the slightest grasp on architectural concerns or separation of responsibilities, it seems that is a perfectly acceptable approach to misuse an API because it conveniently does something close enough to what they want it to do.  There is a reason every serious platform has a separate validation API not coupled to the workflow of submitting an HTML form, and I highly suggest you all find a "smart" PHP developer to create one for Joomla, since this community seems to believe it must own everything itself and has a hard time integrating anything third party, such as the validation components of a tried and true framework like Laravel or Symfony or Zend.

**Comment by HLeithner** [comment]
> > So you say its better to make it impossible to submit an empty string? Sorry that's nonsense.
> 
> If a field is required, generally that means it must have a non-empty value, therefore submitting the field with an empty string should throw a validation error. Assuming you are able to misuse the required attribute of a form field, which has a semantic definition per the HTML standard, to communicate "this field must be submitted as part of the POST, but if it's empty that's perfectly OK" just proves you do not understand basic concepts of data validation.

This it out of context, my concern is about the what brian wrote. remove required from form field if we have a default value, as he wrote I understand it that it's true for all fields or fields with even an empty option field.

And in this case you can't remove required, also formfield makes server side validation to have a consistent description of the form and input data on one place.

**Comment by mbabker** [comment]
> remove required from form field if we have a default value, as he wrote I understand it that it's true for all fields or fields with even an empty option field

The handling of this default value points out yet another critical architecture flaw in the Form API and the MVC interaction with said API.  Ignoring the fact that everything that is in the `Joomla\CMS\Form` namespace today and everything outside of it that inherits from it should just be flat out deprecated and replaced because the entirety of the API is just so frickin' flawed, assigning a default value if a field is empty is not a step that belongs as part of the `Form::filter()` call; in the case of submissions it is a data validation step that should occur either before or after the submitted data is validated (you can decide what is contextually appropriate for your workflow, but it should most definitely be after filtering) and in the case of rendering a field that should be handled when figuring out what value to put into the field at runtime.

A required attribute is not required on a select list with no empty value or on a radio option where it is impossible to not submit that field with something chosen (generally because something is already pre-selected).  You use other mechanisms of validation to ensure a value is set and is in the expected list, you don't misuse the required attribute.  If your excuse is "but you need to assign a value to the field if you're submitting in other ways like a webservices API", then you're already screwing up by trying to re-use a form and data definition not suited for REST oriented contexts.

If everyone's ready to get off their high horses and stop ignoring things because the source is either @brianteeman or @mbabker this pull request is 100% architecturally correct and every response given as to why this pull request is wrong just demonstrates why the Joomla architecture is flawed.  Fix your shitty code.

**Comment by brianteeman** [comment]
Thank you @mbabker I knew I wasn't talking rubbish



> I stand by my comments

**Comment by brianteeman** [comment]
Your arguments for not removing the useless required field when there is a default value _might_ be valid if it was consistently applied. By your arguments **every** field must be a required field. They're not because it would be a silly thing to do. 

Why do you insist that the workflow default field must be required
https://github.com/joomla/joomla-cms/blob/42d9a93b3cc13dd37fd03d9152e3a0e80a402237/administrator/components/com_workflow/forms/workflow.xml#L49
When the equivalent field for a menu item is not
https://github.com/joomla/joomla-cms/blob/42d9a93b3cc13dd37fd03d9152e3a0e80a402237/administrator/components/com_menus/forms/item.xml#L163

Why do you insist that the workflow published field must be required
https://github.com/joomla/joomla-cms/blob/42d9a93b3cc13dd37fd03d9152e3a0e80a402237/administrator/components/com_workflow/forms/workflow.xml#L35

When no other published field is required
https://github.com/joomla/joomla-cms/blob/42d9a93b3cc13dd37fd03d9152e3a0e80a402237/administrator/components/com_banners/forms/banner.xml#L42

Finally making a field required in the form does not ensure that the data submitted is valid.

**Pull request closed without merging**